numba/targets/hashing.py

```diff
@@ -0,0 +1,605 @@
+        # a different (!) Py_hash_t for small string optimization
+        ('djbx33a', DJBX33A),
+        ('expat', EXPAT),
+    ]
+
+
+# Only a few members are needed at present
+_Py_hashSecret = _Py_HashSecret_t.in_dll(pythonapi, '_Py_HashSecret')
+_Py_HashSecret_djbx33a = _Py_hashSecret.djbx33a
+_Py_HashSecret_djbx33a_suffix = _Py_hashSecret.djbx33a.suffix
+_Py_HashSecret_siphash_k0 = _Py_hashSecret.siphash.k0
+_Py_HashSecret_siphash_k1 = _Py_hashSecret.siphash.k1
+# ------------------------------------------------------------------------------
+
+
+@intrinsic
+def grabbyte(typingctx, data, offset):
+    # returns a byte at a given offset in data
+    def impl(context, builder, signature, args):
+        data, idx = args
+        ptr = builder.bitcast(data, ir.IntType(8).as_pointer())
+        ch = builder.load(builder.gep(ptr, [idx]))
+        return ch
+
+    sig = types.uint8(types.voidptr, types.intp)
+    return sig, impl
+
+
+if _Py_hashfunc_name == 'siphash24':
+    # This is a translation of CPython's siphash24 function:
+    # https://github.com/python/cpython/blob/d1dd6be613381b996b9071443ef081de8e5f3aff/Python/pyhash.c#L287-L413
+
+    # /* *********************************************************************
+    # <MIT License>
+    # Copyright (c) 2013  Marek Majkowski <marek@popcount.org>
+
+    # Permission is hereby granted, free of charge, to any person obtaining a copy
+    # of this software and associated documentation files (the "Software"), to deal
+    # in the Software without restriction, including without limitation the rights
+    # to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+    # copies of the Software, and to permit persons to whom the Software is
+    # furnished to do so, subject to the following conditions:
+
+    # The above copyright notice and this permission notice shall be included in
+    # all copies or substantial portions of the Software.
+
+    # THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    # IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    # FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+    # AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    # LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+    # THE SOFTWARE.
+    # </MIT License>
+
+    # Original location:
+    # https://github.com/majek/csiphash/
+
+    # Solution inspired by code from:
+    # Samuel Neves (supercop/crypto_auth/siphash24/little)
+    #djb (supercop/crypto_auth/siphash24/little2)
+    # Jean-Philippe Aumasson (https://131002.net/siphash/siphash24.c)
+
+    # Modified for Python by Christian Heimes:
+    # - C89 / MSVC compatibility
+    # - _rotl64() on Windows
+    # - letoh64() fallback
+    # */
+
+    @intrinsic
+    def grab_uint64_t(typingctx, data, offset):
```

Comment:
Same as above.  Should add to `numba.unsafe`.

Reply:
done.